### ISSUE TITLE
Unify mouse and keyboard item focus/highlight

### ISSUE THREAD

**Issue by jounik** [vaadin/vaadin-combo-box#89]
Currently the user can have two separate highlighted rows in the dropdown: one for the mouse cursor and one for the keyboard.

These two should be one and the same, and the mouse cursor should move the keyboard focus to the highlighted item.

When the mouse cursor leaves the dropdown, the keyboard focus should be cleared, unless there is a previous value/selection in the input. If there is a previous value/selection, the focus/highlight should return to that item.

**Comment by Saulis** [comment]
Things to consider when estimating:
- Currently, changing focused item will trigger it to be scrolled into view (we most likely don't that to happen when using mouse)
- Mouse event listeners need to be added to all item elements
- Not sure how smooth the behaviour is when cursor sits on top of the dropdown while navigating with keyboard if they both change the focused item. (mouse cursor might even prevent you from navigating with the keyboard)

I'm slightly worried about the ROI of the used effort and added complexity - we could have a short discussion in the grooming on how common the targeted use case of using both mouse and keyboard for navigation is.
 
As food for thought, I'd like to propose a visual UX improvement instead, e.g. using border highlighting instead of background highlight for the items focused with a keyboard.

**Comment by jounik** [comment]
Thanks, good points! And sure, this should be discussed before jumping to implementation.

As a reference, seems to work fine for the current combo box: http://demo.vaadin.com/sampler/#ui/data-input/multiple-value/combo-box

I think mouse hover isn’t triggered if an element moves under the mouse without the mouse actually moving. But the scrolling should be prevented when you hover over the items, that’s true.

Why do we need mouse event listeners on all items?

**Comment by Saulis** [comment]
>> Why do we need mouse event listeners on all items?

Well.. don't we need to listen to mouse over on the item elements in order to trigger the necessary changes and figure out which item the mouse is actually hovering on? Or is there some trick to use the CSS `:hover` ?

**Comment by jounik** [comment]
I mean that we only need a mousemove listener on the dropdown element, not all of the individual item elements in the dropdown. Nothing to do with CSS, just minimizing the amount of event listeners needed.

**Comment by Saulis** [comment]
Well.. thinking that way it might be difficult to dig up the index of the individual items - we would probably need to expose the item indexes in the DOM then.

**Comment by jounik** [comment]
Hmm, right, because of iron-list. Perhaps there’s a method in iron-list that maps elements to indexes?

**Comment by Saulis** [comment]
It seems to have something called `modelForElement(el)` which might work - at least combined with some private API from `<iron-list>`

**Comment by jounik** [comment]
1d prototyping research.

**Comment by jounik** [comment]
I think we need to rethink this a little now that we have the value pre-filling implemented (when you navigate with arrow keys the value is pre-filled in the input). Should the same thing happen when the user hover overs items? Probably yes, if the keyboard focus follows the mouse focus.

Trying to play that in my head, it seems like it might work, though I’m fearing users will find it a little odd.

**Comment by jounik** [comment]
Another option for this issue is to visually separate the mouse focus from the keyboard focus (#142).

**Comment by jouni** [comment]
> I think we need to rethink this a little now that we have the value pre-filling implemented (when you navigate with arrow keys the value is pre-filled in the input). Should the same thing happen when the user hover overs items? Probably yes, if the keyboard focus follows the mouse focus.

The current FW8 ComboBox doesn’t actually pre-fill the value when focusing items with the mouse. I think we should follow that design: https://demo.vaadin.com/sampler/#ui/data-input/multiple-value/combo-box
